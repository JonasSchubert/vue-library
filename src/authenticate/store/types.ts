export const ActionTypes = {
  login: 'login',
  logout: 'logout',
  validateToken: 'validate-token'
};

export const GetterTypes = {
  error: 'error',
  isAdministrator: 'isAdministrator',
  isLoading: 'isLoading',
  isUser: 'isUser',
  loggedIn: 'loggedIn',
  token: 'token',
  userRoles: 'userRoles'
};

export const ModuleType = 'authenticate';

export const MutationTypes = {
  setError: 'setError',
  setIsLoading: 'setIsLoading',
  setToken: 'setToken',
  setUserRoles: 'setUserRoles'
};

export const RouteTypes = {
  authenticate: 'authenticate'
};
