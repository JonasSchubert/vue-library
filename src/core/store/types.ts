export const GetterTypes = {
  isAnythingLoading: 'isAnythingLoading'
};

export const MutationTypes = {
  decreaseRequests: 'decreaseRequests',
  increaseRequests: 'increaseRequests'
};
