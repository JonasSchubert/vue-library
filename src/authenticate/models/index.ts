export * from './authenticate-module-setup.d';
export * from './authenticate-state.d';
export * from './base-authenticate-state.d';
export * from './login-data.d';
export * from './login-response.d';
export * from './validate-token-data.d';
