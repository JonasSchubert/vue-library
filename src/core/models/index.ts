export * from './base-state.d';
export * from './entity.d';
