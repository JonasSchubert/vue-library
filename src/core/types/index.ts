export * from './ga-colors.type';
