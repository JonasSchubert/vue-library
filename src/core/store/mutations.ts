import { MutationTree } from 'vuex';
import { BaseState } from '../models';
import { MutationTypes } from './types';

export const createMutations = <T extends BaseState>(): MutationTree<T> => ({
  [MutationTypes.decreaseRequests]: (state: T): void => { state.requests -= 1; },
  [MutationTypes.increaseRequests]: (state: T): void => { state.requests += 1; }
});
