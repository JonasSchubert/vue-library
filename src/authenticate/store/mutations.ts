import { MutationTree } from 'vuex';
import { MutationTypes } from './types';
import { AuthenticateModuleSetup, AuthenticateState, LoginResponse } from '../models';
import { writeCookie, deleteCookie } from '../../core/cookie';

export const createMutations = <T extends LoginResponse>({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup): MutationTree<AuthenticateState<T>> => ({
  [MutationTypes.setData]: (state: AuthenticateState<T>, { data, saveLoginDataTemporary }: { data: T | undefined; saveLoginDataTemporary: boolean }): void => {
    state.data = data;

    if (data?.token !== undefined && saveLoginDataTemporary) {
      writeCookie(cookieKeyAuthenticationToken, data.token, daysTilExpiredAuthenticationCookie);
    } else if (!!data && data.token === undefined) {
      deleteCookie(cookieKeyAuthenticationToken);
    }
  },
  [MutationTypes.setError]: (state: AuthenticateState<T>, { error }: { error: Error | undefined }): void => {
    state.error = error;
  },
  [MutationTypes.setIsLoading]: (state: AuthenticateState<T>, { isLoading }: { isLoading: boolean }): void => {
    state.isLoading = isLoading;
  }
});
