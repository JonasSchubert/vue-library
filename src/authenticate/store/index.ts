import { Module } from 'vuex';
import { AuthenticateModuleSetup, AuthenticateState, LoginResponse } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createAuthenticateState } from './state';

export const createAuthenticateModule = <TRootState, T extends LoginResponse>({ cookieKeyAuthenticationToken, customGetters, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup<TRootState, T>): Module<AuthenticateState<T>, TRootState> => ({
  namespaced: true,
  actions: createActions(),
  getters: createGetters(customGetters),
  mutations: createMutations({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }),
  state: createAuthenticateState()
});
