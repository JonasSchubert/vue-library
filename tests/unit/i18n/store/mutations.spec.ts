import { IVueI18n, LocaleMessageObject, LocaleMessages } from 'vue-i18n';
import { MutationTree } from 'vuex';
import { I18nModuleConfig, I18nState } from '@/i18n/models';
import { createMutations } from '@/i18n/store/mutations';
import { MutationTypes } from '@/i18n/store/types';

describe('createMutations', () => {
  let i18n: IVueI18n;
  let mutations: MutationTree<I18nState>;
  let state: I18nState;

  const availableLocales: string[] = ['de-DE', 'en-GB'];
  const currentLocale = 'en-GB';
  const locales: LocaleMessages = {
    'de-DE': {
      'message': {
        'value': 'Wert'
      }
    },
    'en-GB': {
      'message': {
        'value': 'Value'
      }
    }
  };

  beforeEach(() => {
    i18n = {
      locale: currentLocale,
      messages: locales,
      setLocaleMessage: jest.fn()
    } as unknown as IVueI18n;

    mutations = createMutations({ i18n } as I18nModuleConfig);

    state = {
      availableLocales,
      currentLocale,
      error: undefined,
      locales
    };
  });

  describe(`${MutationTypes.setAvailableLocales}`, () => {
    test('should set available locales', () => {
      // Arrange
      const newAvailableLocales = ['de-DE', 'en-GB', 'es-ES'];

      // Act
      mutations[MutationTypes.setAvailableLocales](state, { availableLocales: newAvailableLocales });

      // Assert
      expect(state.availableLocales).toMatchSnapshot(['de-DE', 'en-GB', 'es-ES']);
    });
  });

  describe(`${MutationTypes.setCurrentLocale}`, () => {
    test('should set current locale', () => {
      // Arrange
      const newCurrentLocale = 'de-DE';

      // Act
      mutations[MutationTypes.setCurrentLocale](state, { currentLocale: newCurrentLocale });

      // Assert
      expect(state.currentLocale).toBe(newCurrentLocale);
      expect(i18n.locale).toBe(newCurrentLocale);
    });
  });

  describe(`${MutationTypes.setError}`, () => {
    test('should set error', () => {
      // Arrange
      const newError: Error = {
        message: 'Message',
        name: 'Name'
      };

      // Act
      mutations[MutationTypes.setError](state, { error: newError });

      // Assert
      expect(state.error).toMatchObject(newError);
    });
  });

  describe(`${MutationTypes.setLocales}`, () => {
    test('should set set locales', () => {
      // Arrange
      const newLocales: LocaleMessages = {
        'de-DE': {
          'message': {
            'test': 'Test',
            'value': 'Wert'
          }
        },
        'en-GB': {
          'message': {
            'test': 'Test',
            'value': 'Value'
          }
        },
        'es-ES': {
          'message': {
            'test': 'Prueba',
            'value': 'valor'
          }
        }
      };

      // Act
      mutations[MutationTypes.setLocales](state, { locales: newLocales });

      // Assert
      expect(state.locales).toMatchObject({
        'de-DE': {
          'message': {
            'test': 'Test',
            'value': 'Wert'
          }
        },
        'en-GB': {
          'message': {
            'test': 'Test',
            'value': 'Value'
          }
        },
        'es-ES': {
          'message': {
            'test': 'Prueba',
            'value': 'valor'
          }
        }
      });
      /* eslint-disable @typescript-eslint/no-explicit-any */
      expect((i18n as any).messages).toMatchObject({
        'de-DE': {
          'message': {
            'test': 'Test',
            'value': 'Wert'
          }
        },
        'en-GB': {
          'message': {
            'test': 'Test',
            'value': 'Value'
          }
        },
        'es-ES': {
          'message': {
            'test': 'Prueba',
            'value': 'valor'
          }
        }
      });
      /* eslint-enable @typescript-eslint/no-explicit-any */
    });
  });

  describe(`${MutationTypes.updateLocales}`, () => {
    test('should set update locales', () => {
      // Arrange
      const locale = 'es-ES';
      const translations: LocaleMessageObject = {
        'message': {
          'value': 'valor'
        }
      };

      // Act
      mutations[MutationTypes.updateLocales](state, { locale, translations });

      // Assert
      expect(state.locales).toMatchObject({
        'de-DE': {
          'message': {
            'value': 'Wert'
          }
        },
        'en-GB': {
          'message': {
            'value': 'Value'
          }
        },
        'es-ES': {
          'message': {
            'value': 'valor'
          }
        }
      });
      /* eslint-disable @typescript-eslint/no-explicit-any */
      expect((i18n as any).setLocaleMessage).toHaveBeenCalledTimes(1);
      expect((i18n as any).setLocaleMessage).toHaveBeenCalledWith('es-ES', {
        'message': {
          'value': 'valor'
        }
      });
      /* eslint-enable @typescript-eslint/no-explicit-any */
    });
  });
});
