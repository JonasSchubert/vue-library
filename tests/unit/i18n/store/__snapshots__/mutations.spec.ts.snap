// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMutations setAvailableLocales should set available locales 1`] = `
Object {
  "0": "de-DE",
  "1": "en-GB",
  "2": "es-ES",
}
`;
